trieste/utils/multi_objectives.py

```diff
@@ -0,0 +1,234 @@
+
+    :param x: The points at which to evaluate the function, with shape [..., d].
+    :param m: The objective numbers.
+    :param k: The input dimensionality for g.
+    :param d: The dimensionality of the synthetic function.
+    :return: The function values at ``x``, with shape [..., m].
+    :raise ValueError (or InvalidArgumentError): If ``x`` has an invalid shape.
+    """
+    tf.debugging.assert_shapes(
+        [(x, (..., d))],
+        message=f"input x dim: {x.shape[-1]} is not align with pre-specified dim: {d}",
+    )
+    tf.debugging.assert_greater(m, 0, message=f"positive objective numbers expected but found {m}")
+
+    def g(xM):
+        return 100 * (
+            k
+            + tf.reduce_sum(
+                (xM - 0.5) ** 2 - tf.cos(20 * math.pi * (xM - 0.5)), axis=-1, keepdims=True
+            )
+        )
+
+    ta = tf.TensorArray(x.dtype, size=m)
+    for i in range(m):
+        xM = x[..., m - 1 :]
+        y = 1 + g(xM)
+        y *= 1 / 2 * tf.reduce_prod(x[..., : m - 1 - i], axis=-1, keepdims=True)
+        if i > 0:
+            y *= 1 - x[..., m - i - 1, tf.newaxis]
+        ta = ta.write(i, y)
+
+    return tf.squeeze(tf.concat(tf.split(ta.stack(), m, axis=0), axis=-1), axis=0)
+
+
+class DTLZ2(DTLZ):
+    """
+    The DTLZ2 problem, the idea pareto fronts lie on (part of) a unit hyper sphere.
+    See :cite:deb2002scalable for details.
+    """
+
+    def objective(self):
+        return partial(dtlz2, m=self.M, d=self.dim)
+
+    def gen_pareto_optimal_points(self, n: int, seed=None):
+        tf.debugging.assert_greater_equal(self.M, 2)
+        rnd = tf.random.normal([n, self.M], seed=seed)
+        samples = tf.abs(rnd / tf.norm(rnd, axis=-1, keepdims=True))
+        return samples
+
+
+def dtlz2(x: TensorType, m: int, d: int) -> TensorType:
+    """
+    The DTLZ2 synthetic function.
+
+    :param x: The points at which to evaluate the function, with shape [..., d].
+    :param m: The objective numbers.
+    :param d: The dimensionality of the synthetic function.
+    :return: The function values at ``x``, with shape [..., m].
+    :raise ValueError (or InvalidArgumentError): If ``x`` has an invalid shape.
+    """
+    tf.debugging.assert_shapes(
+        [(x, (..., d))],
+        message=f"input x dim: {x.shape[-1]} is not align with pre-specified dim: {d}",
+    )
+    tf.debugging.assert_greater(m, 0, message=f"positive objective numbers expected but found {m}")
+
+    def g(xM):
+        z = (xM - 0.5) ** 2
+        return tf.reduce_sum(z, axis=-1, keepdims=True)
+
+    ta = tf.TensorArray(x.dtype, size=m)
+    for i in tf.range(m):
+        y = 1 + g(x[..., m - 1 :])
+        for j in tf.range(m - 1 - i):
+            y *= tf.cos(math.pi / 2 * x[..., j, tf.newaxis])
+        if i > 0:
+            y *= tf.sin(math.pi / 2 * x[..., m - 1 - i, tf.newaxis])
+        ta = ta.write(i, y)
+
+    return tf.squeeze(tf.concat(tf.split(ta.stack(), m, axis=0), axis=-1), axis=0)
```

Comment:
The list has now has been replaced by `tf.TensorArray` as suggested, perhaps could bother @uri-granta for a final look.

Reply:
Looks good to me, though since I'm new to tensorflow tensor manipulation still makes my head hurt. @henrymoss does the conversion from an array of tensors to a concatenated tensor look ok to you?

Reply:
I think so. Can't think of a neater way. 

@TsingQAQ  get this up to date, then we can merge :) I can review your Notebook tomorrow morning.

Reply:
Updated, could we merge it and move forward?